nimbus/src/enrollment.rs

```diff
@@ -20,28 +24,367 @@ pub enum EnrolledReason {
+        })
+    }
+
+    /// Transition our enrollment to WasEnrolled (Option::Some) or delete it (Option::None)
+    /// after an experiment has disappeared from the server.
+    ///
+    /// If we transitioned to WasEnrolled, our enrollment will be garbage collected
+    /// from the database after `PREVIOUS_ENROLLMENTS_GC_TIME`.
+    fn on_experiment_ended(
+        &self,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Option<Self> {
+        log::debug!(
+            "Experiment '{}' vanished while we had enrollment status of {:?}",
+            self.slug,
+            self
+        );
+        let (branch, enrollment_id) = match self.status {
+            EnrollmentStatus::Enrolled {
+                ref branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            EnrollmentStatus::Disqualified {
+                ref branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            _ => return None, // We were never enrolled anyway, simply the enrollment record from the DB.
+        };
+        enrollment_events.push(EnrollmentChangeEvent::new(
+            &self.slug,
+            &enrollment_id,
+            &branch,
+            None,
+            EnrollmentChangeEventType::Unenrollment,
+        ));
+        Some(Self {
+            slug: self.slug.clone(),
+            status: EnrollmentStatus::WasEnrolled {
+                enrollment_id,
+                branch: branch.to_owned(),
+                experiment_ended_at: now_secs(),
+            },
+        })
+    }
+
+    /// Force unenroll ourselves from an experiment.
+    fn on_explicit_opt_out(
+        &self,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Self> {
+        Ok(match self.status {
+            EnrollmentStatus::Enrolled {
+                enrollment_id,
+                ref branch,
+                ..
+            } => {
+                enrollment_events.push(EnrollmentChangeEvent::new(
+                    &self.slug,
+                    &enrollment_id,
+                    &branch,
+                    Some("optout"),
+                    EnrollmentChangeEventType::Disqualification,
+                ));
+                Self {
+                    slug: self.slug.to_string(),
+                    status: EnrollmentStatus::Disqualified {
+                        reason: DisqualifiedReason::OptOut,
+                        branch: branch.to_owned(),
+                        enrollment_id,
+                    },
+                }
+            }
+            EnrollmentStatus::NotEnrolled { .. } => Self {
+                slug: self.slug.to_string(),
+                status: EnrollmentStatus::NotEnrolled {
+                    reason: NotEnrolledReason::OptOut, // Explicitly set the reason to OptOut.
+                },
+            },
+            _ => {
```

Comment:
A small future-looking nit: would it be worth listing the rest of the states explicitly here rather than using a wildcard `_` match? That way if we add new states in future, the compiler will force us to consider how each new state should handle each of these potential transitions.